File: _backend/pkg/handlers/web/deadClick.go
package web

import (
	. "openreplay/backend/pkg/messages"
)

const ClickRelationTime = 1234

type DeadClickDetector struct {
	lastMouseClick     *MouseClick
	lastTimestamp      uint64
	lastClickTimestamp uint64
	lastMessageID      uint64
	inputIDSet         map[uint64]bool
}

func NewDeadClickDetector() *DeadClickDetector {
	return &DeadClickDetector{inputIDSet: make(map[uint64]bool)}
}

func (d *DeadClickDetector) addInputID(id uint64) {
	d.inputIDSet[id] = true
}

func (d *DeadClickDetector) clearInputIDs() {
	d.inputIDSet = make(map[uint64]bool)
}

func (d *DeadClickDetector) reset() {
	d.lastMouseClick = nil
	d.lastClickTimestamp = 0
	d.lastMessageID = 0
	d.clearInputIDs()
}

func (d *DeadClickDetector) Build() Message {
	// remove reset from external Build call
	defer d.reset()
	if d.lastMouseClick == nil || d.lastClickTimestamp+ClickRelationTime > d.lastTimestamp { // reaction is instant
		return nil
	}
	event := &IssueEvent{
		Type:          "dead_click",
		ContextString: d.lastMouseClick.Label,
		Timestamp:     d.lastClickTimestamp,
		MessageID:     d.lastMessageID,
		Context:       d.lastMouseClick.Selector, // hack to pass selector to db (tags filter)
	}
	return event
}

func (d *DeadClickDetector) Handle(message Message, timestamp uint64) Message {
	d.lastTimestamp = timestamp
	switch msg := message.(type) {
	case *SetInputTarget:
		d.addInputID(msg.ID)
	case *CreateDocument:
		d.clearInputIDs()
	case *MouseClick:
		if msg.Label == "" {
			return nil
		}
		isInputEvent := d.inputIDSet[msg.ID]
		event := d.Build()
		if isInputEvent {
			return event
		}
		d.lastMouseClick = msg
		d.lastClickTimestamp = timestamp
		d.lastMessageID = message.MsgID()
		return event
	case *SetNodeAttribute,
		*RemoveNodeAttribute,
		*CreateElementNode,
		*CreateTextNode,
		*SetNodeFocus,
		*MoveNode,
		*RemoveNode,
		*SetCSSData,
		*CSSInsertRule,
		*CSSDeleteRule,
		*SetInputValue,
		*SetInputChecked:
		return d.Build()
	}
	return nil
}

File: tracker/tracker/src/main/detectors/deadClick.ts
import Message, {
  Type,
  MouseClick,
  SetInputTarget,
} from '../../common/messages.gen.js'
import type { Detector, ReportIssue } from './types.js'

/**
 * Port of backend/pkg/handlers/web/deadClick.go.
 *
 * A click is "dead" if the page produced no DOM reaction within
 * CLICK_RELATION_TIME ms of it. Clicks on inputs are ignored (they're expected
 * to just focus). The check is flushed on the next DOM mutation or the next
 * click.
 */

const CLICK_RELATION_TIME = 1234

/**
 * DOM-mutation message types. Any of these arriving means the page reacted to
 * the click. Mirrors the mutation set in the Go Handle() switch
 * (MsgSetCSSData -> SetCSSDataURLBased on the tracker).
 */
const DOM_MUTATION_TYPES: ReadonlySet<number> = new Set<number>([
  Type.SetNodeAttribute,
  Type.RemoveNodeAttribute,
  Type.CreateElementNode,
  Type.CreateTextNode,
  Type.SetNodeFocus,
  Type.MoveNode,
  Type.RemoveNode,
  Type.SetCSSDataURLBased,
  Type.SetInputValue,
  Type.SetInputChecked,
])

export default class DeadClickDetector implements Detector {
  private lastMouseClick: MouseClick | null = null
  private lastTimestamp = 0
  private lastClickTimestamp = 0
  private readonly inputIDSet = new Set<number>()

  constructor(private readonly report: ReportIssue) {}

  private reset(): void {
    this.lastMouseClick = null
    this.lastClickTimestamp = 0
    this.inputIDSet.clear()
  }

  private build(): void {
    const click = this.lastMouseClick
    const clickTs = this.lastClickTimestamp
    const lastTs = this.lastTimestamp
    // matches Go `defer d.reset()`
    this.reset()
    if (click === null || clickTs + CLICK_RELATION_TIME > lastTs) {
      // reaction was instant (or nothing pending) -> not a dead click
      return
    }
    this.report({
      type: 'dead_click',
      contextString: click[3], // label
      context: click[4], // selector (used by the tags filter)
      timestamp: clickTs,
    })
  }

  handle(message: Message, timestamp: number): void {
    const type = message[0]
    const isMutation = DOM_MUTATION_TYPES.has(type)
    if (
      type !== Type.MouseClick &&
      type !== Type.SetInputTarget &&
      type !== Type.CreateDocument &&
      !isMutation
    ) {
      return
    }
    // Go updates lastTimestamp at the top of Handle for every relevant message.
    this.lastTimestamp = timestamp

    switch (type) {
      case Type.MouseClick: {
        const msg = message as MouseClick
        const label = msg[3]
        if (label === '') {
          return
        }
        const isInputEvent = this.inputIDSet.has(msg[1])
        this.build()
        if (isInputEvent) {
          return
        }
        this.lastMouseClick = msg
        this.lastClickTimestamp = timestamp
        return
      }
      case Type.SetInputTarget: {
        const msg = message as SetInputTarget
        this.inputIDSet.add(msg[1])
        return
      }
      case Type.CreateDocument:
        this.inputIDSet.clear()
        return
      default:
        // any DOM mutation flushes the pending click check
        this.build()
    }
  }
}
